Exposed FileLocationSettings in the Editor

RhodeIfado opened the pull request:
I wanted to create a pull request to open the discussion on my changes. 

Making FileLocationSettings non-static might be a breaking change for some users. If they used these static references in their own scripts, for example when you create additional log files or save any raw data from your experiment, they wont be able to access the static references in the same way as before.

I am unsure if this is actually a breaking change, and if so wether it's even worth it trying to implement the feature. It is essential to make FileLocationSettings non-static, to enable serialization.

RhodeIfado committed:
Exposed FileLocationSettings in the Editor

AdamBebko commented on BML_TUX_Project/Packages/bmlTUX/Scripts/Settings/FileLocationSettings.cs:
```diff
@@ -3,48 +3,52 @@
-		public static string SessionLogFilePath => Path.Combine(SessionFolder, SessionLogFileName);
-	}
+  [Serializable]
+	public class FileLocationSettings  {
```
I feel like this class should be a scriptable object. There could be a default version of it with the correct values in the package folder, that is automatically referenced by the design file. The user can create their own version of the scriptable object in their own project and override it if needed (drag theirs in to replace the default reference).

RhodeIfado replied:
I agree this feels right, but this is more difficult than I expected it to be without breaking anything, because any existing Experiment Settings files will become outdated and have to be updated with a reference to the default version. I am unsure how to progress here.

I considered using another script to check for this missing reference in any existing Design / Experiment Settings asset in the project, something like another AssetPostProcess Script filtering the AssetDatabase, but that would trigger on every AssetPostProcess, not just after updating the package. Also that just seems like a hack to avoid a breaking change.

Alternatively, ExperimentDesignFile2 has a Validator that could throw a MissingReference, but I think user will be confused at first where to find the default version if its located in the package, or might not get the idea of creating a new ScriptableObject immediatly.

As far as I understand, other references to Scriptable Objects are setup during initial experiment creation, when utilizing ScriptHelperEditorWindow. Right now I cant think of a way to update existing Designs with a default FileLocationSettings that feels satisfactory. Do you have a good idea?

AdamBebko replied:
You can set a default serialized field's asset reference (Prefab or scriptable object) in the _**Script's**_ inspector. 

So :
1. Create the new file settings default scriptable object. Put it in a good spot in the BMLtux package. (not in the assets folder)
2. Add the serialized field to the experiment design file script. 
3. Find the experiment design file SCRIPT (not an instance of it) in the project view. Click on it and take a look at its inspector. Drag the new file settings SO into the field. 
4. This will ONLY apply on the very first serialization of this field. So in your local project, you'll see nothing happen because you already serialized the field in between steps 2 and 3.
5. But you can test if it worked by creating a new design file and double check the reference is there. Now you can hook it up in your project's design file manually.

![image](https://github.com/BioMotionLab/TUX/assets/43276915/7e20473d-d2c2-4a53-a28f-5b071efd4363)

Let me know you have any questions, this is a pretty obscure Unity feature not too many people know about.

RhodeIfado replied:
Thanks for your great advice! I did not know this feature existed.

Testing this, I wasnt able to validate 6. 
It's possible I am not testing it correctly though. I updated a test project by copying a local version of the package and reimported everything. Not sure if it'll behave differently if it's getting an update from the package manager.

RhodeIfado replied:
Changed in: 5a5212136137a1548ee3cc3f043330f7d107c73b

RhodeIfado replied:
My Project is on 2021.3.23f1

AdamBebko commented on BML_TUX_Project/Packages/bmlTUX/Scripts/Settings/FileLocationSettings.cs:
```diff
@@ -3,48 +3,52 @@
+  [Serializable]
+	public class FileLocationSettings  {
+        [Tooltip("Not every device supports absolute paths. Only advised for Builds on Windows and Mac.\nIf unchecked, Document Folder is used as root.")]
+        public bool dataFolderAsAbsolutePath = false;
```
I think we can let the user figure this out. If they put something different other than the default, then the mistake is on them. I suggest this bool is not needed.

RhodeIfado replied:
Removed in: c5ea5ab6aa8e7fbe4be01de182e5a375055b5c24

AdamBebko commented on BML_TUX_Project/Packages/bmlTUX/Scripts/UI/RuntimeUI/ExperimentGui.cs:
```diff
@@ -119,7 +119,8 @@ public void StartExperimentFromButton()
         public void StartDebugExperimentFromButton() {
-            Session session = new DebugSession();
+            Session session = new DebugSession(runner.DesignFile.GetFileLocationSettings.DebugFolder,
+                runner.DesignFile.GetFileLocationSettings.DebugFileName);
```
The fact that both parameters need to reference `runner.DesignFile.GetFileLocationSettings`, suggests to me that we might want to pass an object around here instead or think of a way to clean this up a bit.

AdamBebko commented on BML_TUX_Project/Packages/bmlTUX/Scripts/Managers/SessionLogger.cs:
```diff
@@ -37,8 +37,8 @@ public static class SessionLogger {
-            if (File.Exists(FileLocationSettings.SessionLogFilePath)) {
-                using (StreamWriter streamWriter = new StreamWriter(FileLocationSettings.SessionLogFilePath, true)) {
+            if (File.Exists(session.saveFilePath)) {
+                using (StreamWriter streamWriter = new StreamWriter(session.saveFilePath, true)) {
```
This is proof that your updates are warranted. The names make way more sense now. nice.

AdamBebko commented on BML_TUX_Project/Packages/bmlTUX/Scripts/ExperimentParts/Session.cs:
```diff
@@ -33,39 +40,39 @@ public class Session {
+            Session session = new Session();
             Debug.Log($"{TuxLog.Prefix} Previous Session file not found, creating new");
-            session = new Session();
+            session.saveFilePath = fileLocationSettings.SessionSaveFilePath;
+            session.sessionFolder = fileLocationSettings.SessionFolder;
```
I don't love this. Because it could introduce bugs when future changes are made if we forget to set one of these fields. Suggest passing the FileLocationSettings reference into the Session constructor.

RhodeIfado replied:
Changed in: 3eb251e42350fa91cf4b8307a77c87236d1cec98

AdamBebko commented on BML_TUX_Project/Packages/bmlTUX/Scripts/ExperimentParts/ExperimentRunner.cs:
```diff
@@ -95,7 +95,7 @@ public abstract class ExperimentRunner : MonoBehaviour {
             DesignFile.GetValidate();
 
-            Session = Session.LoadSessionData();
+            Session = Session.LoadSessionData(DesignFile.GetFileLocationSettings);
```
I don't think we need GetFileLocationSettings anymore, we can just reference the FileLocationSettings scriptable object directly (see below)

RhodeIfado replied:
Not quite sure what you mean, I believe this is still needed?

AdamBebko commented on BML_TUX_Project/Packages/bmlTUX/Editor/PackageInstallDetector.cs:
```diff
@@ -55,8 +55,7 @@ static Version GetPackageVersion()
         static string ProjectFolder => Path.GetDirectoryName(Path.Combine( Application.dataPath, "../" ));
-        static string InstallDataFilePath => Path.Combine(FileLocationSettings.BaseTuxDocumentsFolderPath, ProjectFolder, InstallDataFileName);
-    
+        static string InstallDataFilePath => Path.Combine(ProjectFolder, InstallDataFileName);
```
With these changes, I think we need to rename the file now, otherwise it might be confusing sitting at the root of the project. I suggest we rename it to `bmlTux_InstallData.json`

AdamBebko reviewed (changes requested):
These changes overall look great. I have some suggestions for a bit of cleanup to the architecture, but other than that looks good. Once the changes come in I'll take another look and test it a bit more deeply. Nice work!

Edit: In regards to breaking changes, I think you've minimized them to a large extent, and it will be easy enough for people to update their code to reference the new location.

RhodeIfado committed:
Renamed installData.json

RhodeIfado committed:
Removed absolutePath bool

RhodeIfado committed:
Avoid passing FLS as strings

RhodeIfado committed:
pass FLS into session constructor

RhodeIfado committed:
changed FileLocationSettings to a scriptable object

AdamBebko commented:
What version of Unity? I’ll try to confirm today. Sent from my iPhoneOn Apr 3, 2024, at 7:57 AM, Jakob Rhode-Brähler ***@***.***> wrote:﻿
@RhodeIfado commented on this pull request.



In BML_TUX_Project/Packages/bmlTUX/Scripts/Settings/FileLocationSettings.cs:
> -
-				return tuxPath;
-			}
-		}
-
-		
-		
-		public static string BaseTuxDocumentsFolderPath => Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.MyDocuments), tuxFolderName);
-		public static string SessionFolder => BaseTuxDocumentsFolderPath;
-		public static string DebugFolder => BaseTuxDocumentsFolderPath;
-		
-		public static string LastSessionSaveFilePath => Path.Combine(SessionFolder, LastSessionSaveFileName);
-		public static string SessionLogFilePath => Path.Combine(SessionFolder, SessionLogFileName);
-	}
+  [Serializable]
+	public class FileLocationSettings  {

Thanks for your great advice! I did not know this feature existed.
Testing this, I wasnt able to validate 6.
It's possible I am not testing it correctly though. I updated a test project by copying a local version of the package and reimported everything. Not sure if it'll behave differently if it's getting an update from the package manager.

—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you commented.Message ID: ***@***.***>

RhodeIfado commented:
I added a fix for a bug I've introduced during session logging (1e284d70e48b8f497580aae6be13d5181ba95596)

RhodeIfado committed:
fixed session logging to wrong directory

RhodeIfado commented:
I just found another bug, but I wont have time to fix it today.
During `LoadSessionData `in `Session.cs`, if an old `lastSession.json` is present without the newly added serialized fields `saveFilePath `and `sessionFolder`, the JsonUtility.FromJson will actually fill it with null. This causes an exception along the way.

This method could create a new Session object every time at the top and then use `FromJsonOverwrite`.

RhodeIfado committed:
catch null values at sesson json deserialize

RhodeIfado committed:
update samples with default FileLocationSettings

RhodeIfado commented:
I've updated the samples to include the default reference to the FileLocationSettings Scriptable Object and added a quick fix regarding my recent comment. This fix doesnt feel like the smoothest way to deal with deserializing null values, but its alright. It's better than the fix mentioned in my previous comment though because its much clearer to read.

RhodeIfado commented:
Have you had a chance to confirm it yet?

AdamBebko committed:
fix: added an updater tool to help users with old settings file update it automatically.

AdamBebko commented:
> Have you had a chance to confirm it yet?

sorry for the delay, I have a new baby.  👶 

I got the same findings as you. I wrote a tool based on your branch to auto update user's settings using a dialog.

Could you give me permission to push a branch to your repo? I can PR to your master branch, you can take a look then we can pull it into bmlTUX.

I had a chance to verify the file settings work. Looks good to me. I made one other small change to capitalize the default settings file name.

Thanks for your patience :D

RhodeIfado commented:
Great to hear that you have become a dad!

I have added you as a collaborator on the fork, I am unsure if that is the correct kind of permission as I've never done anything like this on github.

AdamBebko commented:
> Great to hear that you have become a dad!
> 
> I have added you as a collaborator on the fork, I am unsure if that is the correct kind of permission as I've never done anything like this on github.

I created a PR for your review: https://github.com/RhodeIfado/TUX/pull/1

Once accepted and pushed to your main branch, it should automatically update your PR to my repo, and  I'll do a couple small final tests, and we can merge.

RhodeIfado committed:
fix: filename capitalization

RhodeIfado committed:
fix: deleted old installData.json

AdamBebko commented:
@RhodeIfado Merged into dev branch. Need to put it through the build pipeline, but should be released shortly. Thanks so much for your contribution. I'm sure it will make customization easier moving forward!

AdamBebko commented:
:tada: This PR is included in version 5.0.0 :tada:

The release is available on [GitHub release](https://github.com/BioMotionLab/TUX/releases/tag/release/5.0.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
